amigocloud/amigocloud.py

```diff
@@ -366,3 +368,122 @@ def start_listening(self, seconds=None):
         """
 
         self.socketio.wait(seconds=seconds)
+
+    def geocode_addresses(self, project_id, dataset_id,
+                          address_field, geo_field, **params):
+        """
+        Geocode addresses in a dataset. The dataset must have a string field
+        with the addresses to geocode and a geometry field for the Geocoding
+        results.
+        :param project_id: Must be a string.
+        :param dataset_id: Must be a string.
+        :param address_field: Name of the address field in the dataset.
+        :param geo_field: Name of the geometry field in the dataset.
+        :param params: Dictionary to restrict the Geocoding response.
+                       For example: {'country':'PE'}
+                       More information:
+                       https://developers.google.com/maps/documentation/geocoding/intro#ComponentFiltering
+        """
+
+        project_url = ("/projects/{project_id}"
+                       ).format(project_id=project_id)
+
+        dataset_url = ("{project_url}/datasets/{dataset_id}"
+                       ).format(project_url=project_url,
+                                dataset_id=dataset_id)
+        dataset = self.get(dataset_url)
+        total = dataset['feature_count']
+
+        print('%d rows to process' % total)
+        print('Estimated time: %.2f hours' % (total * 2.0 / 3600))
+        start = datetime.now()
+        print('Started at %s hrs' % start.strftime('%H:%M'))
+
+        query_url = ("{project_url}/sql?query={query}&offset={offset}"
+                     "&limit={limit}")
+        query = ("SELECT {address_column}, amigo_id "
+                 "FROM dataset_{dataset_id}"
+                 ).format(address_column=address_field,
+                          dataset_id=dataset_id)
+        rows = []
+
+        print('Exporting addresses...')
+
+        for i in range(0, total, 1000):
+            response = self.get(query_url.format(
+                project_url=project_url,
+                query=query,
+                offset=str(i),
+                limit=1000
+            ))
+            data = response['data']
+            rows.extend(data)
+
+        print('Done!')
+        print('Start processing...')
+
+        geocoder_url = ("/me/geocoder/search?focus.point.lat=0"
+                        "&focus.point.lon=0&text={address}{components}")
+        query_url = "{project_url}/sql"
+
+        components = ''
+        if params:
+            components = '&components='
+            for key, value in params.items():
+                components += ("{key}:{value}"
+                               ).format(key=key, value=value)
+
+        processed = 0
+        points = 0
+        values = ''
+        for row in rows:
```

Comment:
can we call the google api using an asynchronous method? currently this will take too long to process.

just keep in mind the google geocoding rate limiting 😄

Reply:
Done. Google geocoding rate limiting is 50 requests per second.